### PR TITLE
Decorations are instructions

### PR DESCRIPTION
**Pull request by tfoleyNV**
I tried to keep this as distinct commits to make the diffs easier, but the third one ended up being pretty big anyway. The commit message there explains most of what is going on, but the high order bit is that `IRDecoration` now inherits from `IRInst`, and an `IRInst` holds a single list of both decorations and children.

### CODE REVIEW

**Comment by tfoleyNV** [comment]
@jsmall-nvidia I've fixed the build errors on this change, so it is ready for review whenever you get a chance.

**Comment by tfoleyNV** [comment]
This PR should fix #743.

File: source/slang/ir-insts.h [hunk 1]
```diff
@@ -112,63 +163,69 @@ struct IRInterpolationModeDecoration : IRDecoration
...
 struct IRNameHintDecoration : IRDecoration
 {
-    enum { kDecorationOp = kIRDecorationOp_NameHint };
+    enum { kOp = kIROp_NameHintDecoration };
+    IR_LEAF_ISA(NameHintDecoration)
 
-    Name* name;
+    IRStringLit* getNameOperand() { return cast<IRStringLit>(getOperand(0)); }
+
+    UnownedStringSlice getName()
```
Comment (jsmall-nvidia) [inline source/slang/ir-insts.h]: Don't know if we want to assert if getNameOperand result is not null.

File: source/slang/ir-serialize.cpp [hunk 2]
```diff
@@ -543,6 +475,12 @@ Result IRSerialWriter::write(IRModule* module, SourceManager* sourceManager, Opt
                         dstInst.m_payload.m_int64 = irConst->value.intVal;
                         break;
                     }
+                    case kIROp_PtrLit:
```
Comment (jsmall-nvidia) [inline source/slang/ir-serialize.cpp]: Surely this is not really serializable?

File: source/slang/ir-serialize.cpp [hunk 3]
```diff
@@ -1714,6 +1377,13 @@ IRDecoration* IRSerialReader::_createDecoration(const Ser::Inst& srcInst)
                         irConst->value.intVal = srcInst.m_payload.m_int64; 
                         break;
                     }
+                    case kIROp_PtrLit:
```
Comment (jsmall-nvidia) [inline source/slang/ir-serialize.cpp]: Again seems unserializable without special casing what is being pointed to.

File: source/slang/ir.cpp [hunk 4]
```diff
@@ -4689,13 +4708,13 @@ namespace Slang
         IRInst*        val,
         String const&   targetName)
```
Comment (jsmall-nvidia) [inline source/slang/ir.cpp]: perhaps should be UnownedStringSlice

File: source/slang/ir-insts.h (hunk 1 above)
Comment (tfoleyNV) [reply source/slang/ir-insts.h]: The `cast` operation that `getNameOperand` is using here is a checked cast that will assert if the object isn't of the correct type, so there should be no way for it to return null. The naming isn't very good, and we may want to have a pass where we systematically rename things with something like `dynCast`, `checkedCast`, etc. to express the intent better.

**Pull request merged**